### PR TITLE
(MODULES-3037) Manage sources

### PR DESCRIPTION
**Pull request by ferventcoder**
Provide the ability manage a source with a custom type and provider.
This allows a source to be ensurable, and also allows for a source to
be manageable in the same converge that Chocolatey is installed
during. This further allows for installing packages from custom
sources all in the same Puppet run.

### CODE REVIEW

File: spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+require 'spec_helper'
+require 'puppet_x/chocolatey/chocolatey_install'
+
+describe 'Chocolatey Common' do
+
+  context 'on Windows', :if => Puppet::Util::Platform.windows? do
+
```
Comment (ferventcoder) [inline spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb]: This is empty right now, but I wanted to get eyes on this so we can get it merged and through soon. So we can pick this up in other tickets or as maint.

**Comment by ferventcoder** [comment]
@puppetlabs/windows please review.

File: lib/puppet_x/chocolatey/chocolatey_common.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      def file_exists?(path)
+        File.exist?(path)
+      end
+      module_function :file_exists?
+
+      def chocolatey_command
+        if Puppet::Util::Platform.windows?
+          chocoInstallPath = PuppetX::Chocolatey::ChocolateyInstall.install_path
+          #default_location = $::choco_installpath || ENV['ALLUSERSPROFILE'] + '\chocolatey'
+          chocopath = ('C:\ProgramData\chocolatey' if file_exists?('C:\ProgramData\chocolatey\bin\choco.exe')) ||
```
Comment (glennsarti) [inline lib/puppet_x/chocolatey/chocolatey_common.rb]: I'm not 100% comfortable with using a hardcoded path for C:\ProgramData.  It tends to be surfaced as `%ALLUSERSPROFILE%` or at least `%SYSTEMDRIVE%\ProgramData`

File: lib/puppet_x/chocolatey/chocolatey_common.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+              (chocoInstallPath if chocoInstallPath && file_exists?("#{chocoInstallPath}\\bin\\choco.exe")) ||
+              (ENV['ChocolateyInstall'] if ENV['ChocolateyInstall'] && file_exists?("#{ENV['ChocolateyInstall']}\\bin\\choco.exe")) ||
```
Comment (glennsarti) [inline lib/puppet_x/chocolatey/chocolatey_common.rb]: Wouldn't the ChocolateyInstall ENV var have the hightest priority and be at the beginning of this OR code?

File: lib/puppet_x/chocolatey/chocolatey_common.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+          chocopath += '\bin\choco.exe'
+        else
+          chocopath = 'choco.exe'
+        end
+
+        chocopath
+      end
+      module_function :chocolatey_command
+
+      def set_env_chocolateyinstall
```
Comment (glennsarti) [inline lib/puppet_x/chocolatey/chocolatey_common.rb]: Seems to be some overlap between this;
https://github.com/puppetlabs/puppetlabs-chocolatey/blob/master/lib/puppet/provider/package/chocolatey.rb#L92-L94
lib/puppet_x/chocolatey/chocolatey_common.rb#L33-36
Unless I'm missing something?

File: spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  context ".choco_exe?" do
+    it "should be true when version is 0.9.9.0" do
+      PuppetX::Chocolatey::ChocolateyCommon.expects(:choco_version).returns('0.9.9.0')
+      PuppetX::Chocolatey::ChocolateyCommon.clear_cached_values
+      PuppetX::Chocolatey::ChocolateyCommon.choco_exe?.must be_truthy
+    end
+
+    it "should be true when version is greater than 0.9.9.0" do
+      PuppetX::Chocolatey::ChocolateyCommon.expects(:choco_version).returns('0.9.9.0')
```
Comment (glennsarti) [inline spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb]: This appears to be a duplicate of the test above.  Is `.returns('0.9.9.0')` correct here?

**Comment by glennsarti** [comment]
Is it worth refactoring out the "magic" version numbers in the unit tests  e.g. 0.9.9.10 etc?

File: lib/puppet_x/chocolatey/chocolatey_common.rb (hunk 4 above)
Comment (ferventcoder) [inline lib/puppet_x/chocolatey/chocolatey_common.rb]: No, not at all. The plan is to convert the package code over to using this. It was getting a little out of scope for this ticket as it was.

File: spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb (hunk 5 above)
Comment (ferventcoder) [inline spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb]: should have been a higher number.

File: lib/puppet_x/chocolatey/chocolatey_common.rb (hunk 2 above)
Comment (ferventcoder) [inline lib/puppet_x/chocolatey/chocolatey_common.rb]: This section is a straight copy from the package provider. There be voodo and reasons there, they haven't been all determined yet.

File: lib/puppet_x/chocolatey/chocolatey_common.rb (hunk 3 above)
Comment (ferventcoder) [inline lib/puppet_x/chocolatey/chocolatey_common.rb]: There is a reason behind the order. I'm just not sure what it is.

Comment (ferventcoder) [inline lib/puppet_x/chocolatey/chocolatey_common.rb]: It does have to do with suitability and first run.

**Comment by ferventcoder** [comment]
> Is it worth refactoring out the "magic" version numbers in the unit tests e.g. 0.9.9.10 etc?

Yeah, I think we should make it clearer - I'll change it to minimum_supported_version - that should make it pretty clear what is being tested.

File: spec/unit/puppet_x/chocolatey/chocolatey_install_spec.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+      expected_value = 'C:\ProgramData\chocolatey'
+      Win32::Registry.any_instance.expects(:[]).with('ChocolateyInstall').raises(Win32::Registry::Error.new(2), 'file not found yo')
+
+      PuppetX::Chocolatey::ChocolateyInstall.install_path.must == expected_value
+    end
+  end
+
+  context 'on Linux', :if => Puppet.features.posix? do
+    it "should return the default install path on a non-windows system" do
+      PuppetX::Chocolatey::ChocolateyInstall.install_path.must == 'C:\ProgramData\chocolatey'
```
Comment (glennsarti) [inline spec/unit/puppet_x/chocolatey/chocolatey_install_spec.rb]: This seems counter-intuitive.  On a linux system wouldn't the install path be nil or empty?

File: spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+describe 'Chocolatey Common' do
+
+  let (:first_compiled_choco_version) {'0.9.9.0'}
+  let (:newer_choco_version) {'0.9.10.0'}
+  let (:last_posh_choco_version) {'0.9.8.33'}
+
+  context ".choco_version" do
+    it "should return PuppetX::Chocolatey::ChocolateyVersion.version" do
+      expected = '0.9.9.0.1'
```
Comment (glennsarti) [inline spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb]: Are you using a 5 segment version here on purpose?  Seems most of the version for chocolatey are 4 segment ala SemVer

File: spec/unit/puppet/type/chocolateysource_spec.rb [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+      it "should accept any string that represents a numeric value" do
+        resource[param_symbol] = '1'
+        resource[param_symbol] = '0'
+      end
+
+      it "should not accept other string values" do
+        expect {
+          resource[param_symbol] = 'value'
+        }.to raise_error(Puppet::Error, /An integer is necessary for priority/)
```
Comment (glennsarti) [inline spec/unit/puppet/type/chocolateysource_spec.rb]: I'm don't think you should have the word `priority` in the regex.  What if you added another numeric parameter?

File: spec/unit/puppet/type/chocolateysource_spec.rb [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+      end
+
+      it "should not accept symbol values" do
+        expect {
+          resource[param_symbol] = :whenever
+        }.to raise_error(Puppet::Error, /An integer is necessary for priority/)
```
Comment (glennsarti) [inline spec/unit/puppet/type/chocolateysource_spec.rb]: I don't think you should have the word `priority` in the regex.  What if you added another numeric parameter?

File: spec/unit/puppet/type/chocolateysource_spec.rb [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    context "parameter :#{param}" do
+      let (:param_symbol) { param.to_sym }
+
+      it "should accept any numeric value" do
+        resource[param_symbol] = 0
+        resource[param_symbol] = 10
+      end
+
+      it "should accept any string that represents a numeric value" do
+        resource[param_symbol] = '1'
```
Comment (glennsarti) [inline spec/unit/puppet/type/chocolateysource_spec.rb]: Do you need to test for negative integers?  e.g. -1

File: spec/unit/puppet_x/chocolatey/chocolatey_install_spec.rb (hunk 6 above)
Comment (ferventcoder) [inline spec/unit/puppet_x/chocolatey/chocolatey_install_spec.rb]: Default path when one is not found.

File: spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb (hunk 7 above)
Comment (ferventcoder) [inline spec/unit/puppet_x/chocolatey/chocolatey_common_spec.rb]: Yes, it's so that the verification is valid. I could accidentally have a false positive on a version if the version was a real one.

File: spec/unit/puppet/type/chocolateysource_spec.rb (hunk 8 above)
Comment (ferventcoder) [inline spec/unit/puppet/type/chocolateysource_spec.rb]: Good point. That was a mistake.

File: spec/unit/puppet/type/chocolateysource_spec.rb (hunk 10 above)
Comment (ferventcoder) [inline spec/unit/puppet/type/chocolateysource_spec.rb]: It will allow it.

File: spec/unit/puppet/type/chocolateysource_spec.rb (hunk 9 above)
Comment (ferventcoder) [inline spec/unit/puppet/type/chocolateysource_spec.rb]: Thanks for noticing that.

File: lib/puppet_x/chocolatey/chocolatey_common.rb (hunk 4 above)
Comment (ferventcoder) [inline lib/puppet_x/chocolatey/chocolatey_common.rb]: No more overlap.

**Comment by ferventcoder** [comment]
@puppetlabs/windows paired with @glennsarti some today on this as we fought through refactoring the Chocolatey package provider bits to get them to use `PuppetX::Chocolatey::ChocolateyCommon`. Lots of tests needed touched and adjusted. 

There may be an underlying issue or two with when choco.exe is actually found (every system that runs these tests has Chocolatey installed). We removed it for a bit to ensure things would work without it.

**Comment by ferventcoder** [comment]
I'll pick these up tomorrow and work to get them overridden.

**Comment by ferventcoder** [comment]
@puppetlabs/windows all comments have been addressed. Ready to go.

**Comment by glennsarti** [comment]
Looks good to me.
